@considerate my test suite has a type error that doesn't make much sense to me. It seems as though a `ByteString` needs to instantiate the `Content` type class in order to be used as the content of a `Ref`. I can't figure out why the type checker thinks this should be the case, and if it is, we definitely need to add the instance for it into the `Content` module.

```
  test/Main.hs:49:18: error:
      • No instance for (Casper.Content ByteString)
          arising from a use of ‘Casper.newRef’
      • In a stmt of a 'do' block: ref <- Casper.newRef hello
        In the second argument of ‘($)’, namely
          ‘do ref <- Casper.newRef hello
              (TrivialTestType refList) <- Casper.readVar root
              Casper.writeVar root $ TrivialTestType $ ref : refList’
        In the second argument of ‘($)’, namely
          ‘Casper.transact
             $ do ref <- Casper.newRef hello
                  (TrivialTestType refList) <- Casper.readVar root
                  Casper.writeVar root $ TrivialTestType $ ref : refList’
     |
  49 |           ref <- Casper.newRef hello
     |                  ^^^^^^^^^^^^^^^^^^^
```

--- a/README.md
+++ b/README.md
@@ -1,13 +1,96 @@
-# casper
+# Casper: a content-addressable store
 
-Extremely simple content-addressable storage.
+`Casper` is a content addresable store using ordinary directories, and
+content-addressible files (files named after the SHA-1 hash of the file's
+conetnt) in the file system as the database, similar to how the Git revision
+control system stores information. All filesystem access through a single
+`Store` handle in a single Haskell process is guaranteed to be thread safe by
+the STM library. However **no guarantees** are provided as to the consistency
+of the database outside of the `Store` handle of a single Haskell process, for
+example, if an application were to create multiple `Store`s to the same
+filesystem directory accessible by multiple threads, or if multiple processes
+were to access a single Casper store.
+
+The `openStore` function is the entry point to the Casper database, which
+evaluates functions of type `CasperT`. The `Store` data type itself, and all
+types stored within (such as `Ref` and `Var`) have an `s` type variable which
+serves the same purpose as the `s` type variable of the `Control.Monad.ST.ST`
+function type: to prevent conflicting stateful updates to the Store content. It
+is expected that a Casper store will be opened with `openStore` in a separate
+process thread only once, at process initialization, and a function in the
+`CasperT` monad will loop forever, listining for requests on an incoming
+channel for database lookups and updates.
+
+To actually store and retrieve data with Casper, start by defining a `root`
+data type that can contain all of the information that is stored in the
+database, such as a list or a tree of immutable `Ref` values or mutable `Var`
+values. Both `Var` and `Ref` data types can themselves be stored into other
+data types that can be written to the store. By traversing through the `root`
+data type, resolving `Ref`s and `Var`s as you go, you can extract pure data
+types from the store You can also fold in the other direction, converting pure
+data types to storable data types.
+
+The `root` data type is stored into a single object in the database, and
+contains an index to all other objects stored in the database. This `root`
+object should be stored into an immutable `Var`, unless the database is to be
+read-only. The `root` data type must take at least one type variable `s` which
+serves a similar purpose to the `s` value of the `Control.Monad.ST.ST` monad.
+
+It also is necessary for the `root` data type to derive several type classes,
+at a minimum:
+
+  - `Content` from `Casper` in this `casper` package
+  - `Generic` from `GHC.Generics` in `base`
+  - `FromJSON` and `ToJSON` from `Data.Aeson` in the `aeson` package
+  - `Serialize` from `Data.Serialize` in the `cereal` package
+
+The easiest way to instantiate these type classes is with the GHC language
+extensions:
+
+  - `DeriveGeneric`
+  - `DeriveAnyClass`
+  - `XDerivingStrategies`
+  - `DerivingVia`
+
+Casper content is stored as `ByteString`s in the database. The `Casper` module
+provides a `WrapAeson` newtype that instantiates `Serialize` in such a way that
+data is serialized as a JSON string without prefixing the string with a binary
+integer string length. See the `WrapAeson` data type for more information.
+
+Here is a template for a program that uses a Casper database:
 
 ```haskell
-main = do
-  store <- initStore "./store"
-  runCasperT $ do
-    addr :: Address [Int] <- write [1,2,3]
-    ...
-    read addr >>= liftIO . print
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DerivingVia #-}
+module Main where
+import GHC.Generics (Generic)
+import Data.Aeson (FromJSON, ToJSON) -- from the "aeson" package
+import Data.Serialize (Serialize) -- from the "cereal" package
 
+data Root s
+    = Root
+      { ...
+      }
+    deriving stock ('Generic')
+    deriving ('Content')
+    deriving anyclass ('FromJSON', 'ToJSON')
+    deriving ('Serialize') via WrapAeson (Root s)
+
+emptyRoot :: Root s
+emptyRoot = Root{ ... }
+
+main :: IO ()
+main = do
+    ...
+    (root, store) <- 'openStore' "/path/to/store" emptyRoot $ \ currentRoot -> do
+        -- store initial values
+        root <- transact $ newVar currentRoot
+        store <- 'getStore'
+        pure (root, store)
+    ...
+    forever $ do
+        ...
+        runCasperT store $ do
+            -- read/write operations on the Casper store
+        ...
 ```

--- a/casper.cabal
+++ b/casper.cabal
@@ -56,14 +56,16 @@ test-suite casper-tests
   default-language: Haskell2010
   build-depends:
       aeson
-    , base              >=4.7      && <5
+    , base
     , bytestring
     , casper
     , cereal
     , containers
-    , directory         >=1.3.6.0  && <1.4
-    , filepath          >=1.4.2.1  && <1.5
-    , mtl               >=2.2.2    && <2.3
-    , tasty             >=1.3.1    && <1.4
-    , tasty-hspec       >=1.1.5.1  && <1.2
+    , directory
+    , filepath
+    , hspec
+    , mtl
+    , tasty
+    , tasty-hspec
+    , temporary
     , tf-random

--- a/src/Casper.hs
+++ b/src/Casper.hs
@@ -13,30 +13,37 @@
 {-# OPTIONS_GHC -Wno-unused-imports #-}
 {-# OPTIONS_GHC -Wno-unused-top-binds #-}
 
--- | "Casper" is a content addresable store using ordinary directories and files in the file system
--- as the database, analogous to how the Git revision control system stores information. All
--- filesystem access through __a single__ 'Store' handle in a single Haskell process is guaranteed
--- to be thread safe by the STM library. However __no guarantees__ are provided as to the
--- consistency of the database outside of the 'Store' handle of a single Haskell process, for
--- example, if an application were to create multiple 'Store's to the same filesystem directory
--- accessible by multiple threads, or if multiple processes were to access a single Casper store.
+-- | "Casper" is a content addresable store using ordinary directories, and content-addressible
+-- files (files named after the SHA-1 hash of the file's conetnt) in the file system as the
+-- database, similar to how the Git revision control system stores information. All filesystem
+-- access through a single 'Store' handle in a single Haskell process is guaranteed to be thread
+-- safe by the STM library. However __no guarantees__ are provided as to the consistency of the
+-- database outside of the 'Store' handle of a single Haskell process, for example, if an
+-- application were to create multiple 'Store's to the same filesystem directory accessible by
+-- multiple threads, or if multiple processes were to access a single Casper store.
 --
--- To use Casper, start by defining a @root@ data type that can contain all of the information that
--- is stored in the database, such as a list or a tree of immutable 'Ref' values or mutable 'Var'
--- values. Both 'Var' and 'Ref' data types can themselves be stored into other data types that can
--- be written to the store. The @root@ data type is stored into a single object in the database, and
--- contains an index to all other objects stored in the database. This @root@ object should be
--- stored into an immutable 'Var', unless the database is to be read-only.
+-- The 'openStore' function is the entry point to the Casper database, which evaluates functions of
+-- type 'CasperT'. The 'Store' data type itself, and all types stored within (such as 'Ref' and
+-- 'Var') have an @s@ type variable which serves the same purpose as the @s@ type variable of the
+-- 'Control.Monad.ST.ST' function type: to prevent conflicting stateful updates to the Store
+-- content. It is expected that a Casper store will be opened with 'openStore' in a separate process
+-- thread only once, at process initialization, and a function in the 'CasperT' monad will loop
+-- forever, listining for requests on an incoming channel for database lookups and updates.
 --
--- When you initialize your program, initialize the Casper database with the 'openStore'
--- function. Assuming you want your root object to be mutable, create a 'Var' with the @root@
--- value. Then use the 'getStore' function to retrieve the handle to the Casper store. After that,
--- you can use 'runCasperT' on the Casper store to read from the database, or to run 'Transaction's
--- that modify the content of the database.
+-- To actually store and retrieve data with Casper, start by defining a @root@ data type that can
+-- contain all of the information that is stored in the database, such as a list or a tree of
+-- immutable 'Ref' values or mutable 'Var' values. Both 'Var' and 'Ref' data types can themselves be
+-- stored into other data types that can be written to the store. By traversing through the @root@
+-- data type, resolving 'Ref's and 'Var's as you go, you can extract pure data types from the store
+-- You can also fold in the other direction, converting pure data types to storable data types.
+-- 
+-- The @root@ data type is stored into a single object in the database, and contains an index to all
+-- other objects stored in the database. This @root@ object should be stored into an immutable
+-- 'Var', unless the database is to be read-only. The "root" data type must take at least one type
+-- variable @s@ which serves a similar purpose to the @s@ value of the 'Control.Monad.ST.ST'
+-- monad.
 --
--- When declaring a "root" data type, it must take at least one type variable @s@ which serves a
--- similar purpose to the @s@ value of the 'Control.Monad.ST.ST' monad. It is also necessary for the
--- "root" data type to derive several type classes, at a minimum:
+-- It also is necessary for the "root" data type to derive several type classes, at a minimum:
 --
 --   - 'Content' from "Casper" in this "casper" package
 --   - 'Generic' from "GHC.Generics" in "base"

--- a/src/Content.hs
+++ b/src/Content.hs
@@ -10,6 +10,8 @@ module Content where
 import GHC.Generics
 import Ref (Ref)
 import Var (Var)
+import qualified Data.ByteString as Strict
+import qualified Data.ByteString.Lazy as Lazy
 
 class Content a where
   refs ::
@@ -50,3 +52,7 @@ instance Content (Var a s) where refs fr _ r = pure $ fr r
 instance Content a => Content [a]
 
 instance Content Int where refs _ _ _ = []
+
+instance Content Strict.ByteString where refs _ _ _ = []
+
+instance Content Lazy.ByteString where refs _ _ _ = []

--- a/test/Main.hs
+++ b/test/Main.hs
@@ -7,18 +7,16 @@
 import qualified Casper
 import Casper (CasperT)
 
-import Control.Monad (unless)
-
 import Data.Aeson (FromJSON, ToJSON)
 import Data.ByteString.Char8 (ByteString)
 import Data.Serialize (Serialize)
 import GHC.Generics (Generic)
 
-import System.Directory
-import System.FilePath
+import System.IO.Temp (withTempDirectory)
 
+import Test.Hspec (Spec, it, example)
 import Test.Tasty (TestTree, defaultMain, testGroup)
-import Test.Tasty.Hspec (Spec, testSpecs, it, example)
+import Test.Tasty.Hspec (testSpecs)
 
 ----------------------------------------------------------------------------------------------------
 
@@ -33,33 +31,18 @@ newtype TrivialTestType s = TrivialTestType [Casper.Ref TestString s]
 initTrivial :: TrivialTestType s
 initTrivial = TrivialTestType []
 
--- | This test suite does not use temporary directories, it stores everything in the same place
--- every time. The reason is so that if anything goes wrong we can inspect the state of the
--- database.
-testStorePath :: FilePath
-testStorePath = "./casper-test-store"
-
--- | Create a new store, and populate it with a single 'Capser.Var' containing a whole entire
--- 'TrivialTestType'.
-trivialTestTypeNewStore
-  :: TrivialTestType s
-  -> CasperT s IO (Casper.Var (TrivialTestType s) s, Casper.Store s)
-trivialTestTypeNewStore root = do
-  root <- Casper.transact $ Casper.newVar root
-  store <- Casper.getStore
-  pure (root, store)
-
 simpleTests :: Spec
-simpleTests = do
-  let hello = "Hello, world!" :: TestString
+simpleTests =
+  let hello = "Hello, world!" :: TestString in
   it ("opens a new content addressable store, stores a single value " <> show hello) $
-    example $ do
-      (root, store) <- Casper.openStore testStorePath initTrivial trivialTestTypeNewStore
-      Casper.runCasperT store $
-        Casper.transact $ do
-          ref <- Casper.newRef hello
-          (TrivialTestType refList) <- Casper.readVar root
-          Casper.writeVar root $ TrivialTestType $ ref:refList
+  example $
+  withTempDirectory "." "casper-test-store.XXXXXXXX" $ \ testStorePath ->
+  Casper.openStore testStorePath initTrivial $ \ refList0 -> do
+    rootVar <- Casper.transact $ Casper.newVar refList0
+    Casper.transact $ do
+      ref <- Casper.newRef hello
+      (TrivialTestType refList) <- Casper.readVar rootVar
+      Casper.writeVar rootVar $ TrivialTestType $ ref:refList
 
 main :: IO ()
 main =

--- a/test/Main.hs
+++ b/test/Main.hs
@@ -5,6 +5,7 @@
 {-# LANGUAGE DerivingVia #-}
 
 import qualified Casper
+import Casper (CasperT)
 
 import Control.Monad (unless)
 
@@ -32,18 +33,28 @@ newtype TrivialTestType s = TrivialTestType [Casper.Ref TestString s]
 initTrivial :: TrivialTestType s
 initTrivial = TrivialTestType []
 
+-- | This test suite does not use temporary directories, it stores everything in the same place
+-- every time. The reason is so that if anything goes wrong we can inspect the state of the
+-- database.
 testStorePath :: FilePath
 testStorePath = "./casper-test-store"
+
+-- | Create a new store, and populate it with a single 'Capser.Var' containing a whole entire
+-- 'TrivialTestType'.
+trivialTestTypeNewStore
+  :: TrivialTestType s
+  -> CasperT s IO (Casper.Var (TrivialTestType s) s, Casper.Store s)
+trivialTestTypeNewStore root = do
+  root <- Casper.transact $ Casper.newVar root
+  store <- Casper.getStore
+  pure (root, store)
 
 simpleTests :: Spec
 simpleTests = do
   let hello = "Hello, world!" :: TestString
   it ("opens a new content addressable store, stores a single value " <> show hello) $
     example $ do
-      (root, store) <- Casper.openStore testStorePath initTrivial $ \ root -> do
-        root <- Casper.transact $ Casper.newVar root
-        store <- Casper.getStore
-        pure (root, store)
+      (root, store) <- Casper.openStore testStorePath initTrivial trivialTestTypeNewStore
       Casper.runCasperT store $
         Casper.transact $ do
           ref <- Casper.newRef hello

